### PR TITLE
tests: isolate shellspec scratch Git config

### LINKED ISSUE #1984: shellspec: scratch-repo git commits inherit the developer's global config (92 failures under commit.gpgsign)
Surfaced by the adversarial review of #1975 (the pytest half of #1967) as an unswept sibling of the same defect class. Reproduced independently before filing.

### The defect

The shellspec suite builds throwaway repositories and commits into them without neutralising the developer's global/system Git configuration — the same defect #1967 reported for the pytest suite. Under the exact condition #1967 names (`commit.gpgsign=true` + `gpg.format=ssh` + an unusable `user.signingkey` in `~/.gitconfig`):

```
CLEAN ~/.gitconfig:        1116 examples, 0 failures, 8 skips
gpgsign ~/.gitconfig:      1116 examples, 92 failures, 8 skips
```

Failure signature is the one the issue reports:

```
fatal: failed to write commit object
```

`shellspec` is a `run-gates.sh` and CI gate, so this fails the canonical gate on any developer machine with global commit signing — exactly the symptom #1967 was filed for, in the other harness.

### Why it is not fixed in #1975

That PR swept the pytest suite (55 failures across six helpers, now sharing one `tests/gitenv.py` definition and pinned by `tests/test_gitenv.py`). The shell suite is a different mechanism — each spec drives `git` directly rather than through a Python helper — so it is a separate change with its own shape, not a line added to the same sweep. #1975's claim was narrowed to the pytest suite rather than left implying a completed class sweep.

### Notes for whoever takes it

- Several specs already carry a point defence (`git config commit.gpgsign false` after init, e.g. `agent_run_gates_spec.sh`, `githooks_pre_push_lease_spec.sh`). That covers signing only; a global `core.hooksPath` that resolves would still make a scratch commit run foreign hooks. Prefer neutralising both config scopes (`GIT_CONFIG_GLOBAL=/dev/null GIT_CONFIG_SYSTEM=/dev/null`) over per-knob defences.
- ADR-47 already requires shellspec git ops to scrub `GIT_*`; this is the config-scope half of the same rule.
- The pytest side's rationale is worth mirroring: the scrub belongs to repository SETUP, and must not silently change the environment a checker-under-test observes.

### Acceptance

- WHEN `~/.gitconfig` sets `commit.gpgsign=true` with an unusable signing key AND `shellspec` runs, THEN the suite reports 0 failures.
- A spec pins the property, so removing a scrub from any swept spec fails the suite on a clean machine too.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by andrebrait**
## Summary

- add a per-call `git_fixture` helper that neutralizes global and system Git configuration without changing the environment seen by scripts under test
- sweep all 14 shellspec scratch-repository fixtures and assertions to use the helper
- extend the existing guard with independently pinned command-position, prefix, quoting, marker, and oversized-line coverage

## Reproduction and verification

On the then-current `origin/devel`, a temporary `HOME/.gitconfig` with `commit.gpgsign=true`, `gpg.format=ssh`, and a missing signing key produced 100 shellspec failures. A focused failure ended with:

```text
fatal: failed to write commit object
```

Final executed evidence:

- hostile signing config across all 14 swept specs: 204 examples, 0 failures
- canonical `scripts/agent/run-gates.sh --diff origin/devel`: `GATES: PASS`
- original frozen proof (`6f15c901…`): 10 examples / 4 failures RED, 10 / 0 GREEN
- scanner supplemental proof (`f7f0b241…`): 19 / 3 RED, 19 / 0 GREEN
- quoted-marker proof (`80f14414…`): 1 / 1 RED, 1 / 0 GREEN
- independent final verification: PASS; all 19 changed files reviewed; no skipped checks

The three similarly named pytest commits (`65d849d0`, `14cc20bb`, and `22a596ae`) intentionally covered only the pytest harness. This PR completes the separate shellspec mechanism.

Fixes #1984

🤖 Generated by OpenAI Codex and posted on behalf of @andrebrait.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved shell safeguards to reliably detect unapproved raw Git fixture commands.
  - Reduced false positives by ignoring comments and quoted text while preserving valid command detection.
  - Same-line exemption markers now require a meaningful reason.
  - Added clearer remediation messages identifying the affected Git operation.
  - Improved handling of complex command syntax, prefixes, and long lines.

- **Tests**
  - Standardized repository fixtures to run with isolated Git configuration.
  - Expanded coverage for command detection, exemptions, quoting, and edge cases.

### CODE REVIEW

**Comment by andrebrait** [comment]
@coderabbitai review

🤖 Generated by OpenAI Codex and posted on behalf of @andrebrait.

File: scripts/git-env-scrub-guard.sh [hunk 1]
```diff
@@ -55,20 +52,140 @@ grep -Ern --include='*.sh' \
...
+    function raw_git(line) {
+        prefix="([[:alnum:]_]+=[^[:space:]]+[[:space:]]+)*"
+        env_prefix="env([[:space:]]+[^;&|[:space:]]+)*[[:space:]]+"
+        boundary="(^[[:space:]]*|[;&|!][[:space:]]*)"
+        sub_boundary="\\$\\([[:space:]]*"
+        return line ~ (boundary prefix "git([[:space:]]|$)") \
+            || line ~ (boundary env_prefix "git([[:space:]]|$)") \
+            || line ~ (sub_boundary prefix "git([[:space:]]|$)") \
+            || line ~ (sub_boundary env_prefix "git([[:space:]]|$)")
+    }
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply scripts/git-env-scrub-guard.sh:201]: Verified: the lexical scanner alone misses valid shell forms. I fixed the invariant at the completed sweep instead of extending a home-grown parser: the guard now pins the exact `git_fixture` occurrence count in each of the 14 swept specs, so removing any fixture pin fails regardless of surrounding shell syntax. A new regression mutates the first pin in every swept spec and is red on the previous head (the three function-body forms escaped) and green on 7d30f709. The scanner remains only a bounded lint for common direct-command mistakes. Hostile `commit.gpgsign=true` run: 204 examples, 0 failures; canonical gates pass.

🤖 Generated by OpenAI Codex and posted on behalf of @andrebrait.

**Comment by andrebrait** [comment]
Review audit for final head `ef391b19495dac03f1cef33c75c34ae4a493276e`:

- Whole-PR adversarial review: `gpt-5.6-sol`, full profile, xhigh effort. Routed top-tier because the PR changes 20 files and 899 lines. Every changed file was examined.
- Review-fix delta re-review: `gpt-5.6-luna`, full profile, xhigh effort, scoped to the fix commits ending at `7d30f709`. Both delta files were examined. Verdict: PASS. The final rebase changed ancestry and commit IDs only; the reviewed two-file patch is unchanged.
- CodeRabbit completed one actionable inline review. Its finding duplicated adversarial finding 1 below and received a threaded reply. The thread is resolved.
- An unsolicited Copilot review appeared through repository automation; it reviewed 19 files and posted no findings. No Copilot review was requested.
- Snyk is absent on this head; advisory scan skipped.

Findings ledger:

1. `scripts/git-env-scrub-guard.sh`: the lexical scanner missed valid shell command forms, including three actual swept helper-removal mutations. `fixed@ef391b19` (rebased from reviewed commit `7d30f709`): the completed-sweep invariant is now an exact per-file `git_fixture` count manifest, independent of shell syntax. Frozen regression `d9a3d5fee7b4edddded57d189b7fa23f9a4f97a4` is RED on the pre-fix head (`1 example, 1 failure`, missed `agent_run_gates_spec.sh`) and GREEN on the reviewed patch (`1 example, 0 failures`); it mutates every one of the 14 swept specs. Combined guard suite: 31 examples, 0 failures.
2. `tests/shell/git_env_scrub_spec.sh`: two descriptions still name `scrub_git_env` although the clause-2 implementation now pins `git_fixture`. `skipped`: non-blocking wording nit in the byte-frozen original red/green proof; changing it would discard the recorded proof hash. Assertions remain correct, and the new manifest spec documents the final invariant precisely.
3. CodeRabbit docstring-coverage warning: `skipped`: generic unenforced metric does not apply to this POSIX-shell change; canonical ShellCheck, ShellSpec, and repository gates pass.

Executed final evidence: hostile global Git config with `commit.gpgsign=true` and a missing SSH signing key passed all 204 swept examples; `scripts/agent/run-gates.sh` passed; exact-head GitHub CI on `ef391b19` passed every required check, including ShellSpec, pytest, ShellCheck, PHPStan, PHPUnit, PHPCS, and the aggregate `All tests passed` gate.

🤖 Generated by OpenAI Codex and posted on behalf of @andrebrait.
